File: salary.go
package main

import "fmt"

// I check for:
// Readability
// Understandability
// Performance
// Error handling
// Bugs & Logical flaws
// DRY-Principle -> Don't write redundant code
// SOP-Principle -> Each thing does its own thing
// KISS-Principle -> Try to keep it simple
// SOLID-Principle -> https://en.wikipedia.org/wiki/SOLID
// POLA-Principle -> Code should work as one would expect
// YAGNI-Principle -> Don't code stuff you might not need

// Added employee interface for
// -> Type safety.
// -> Polymorphism and modularity.
type EmployeeInterface interface {
	getId() int
	getName() string
}

// Added salary interface because salary class violated several SOLID-Principles
// -> Single responsibility principle. Salary function did not only calculate but also check what type of employee it was
// -> Open/closed principle. Salary function was not open for extension. If we wanted to add a new employee we would have to change the salary function
type SalaryInterface interface {
	calculateSalary() float64
}

// implemented employee interface
// Removed position property because it was redundant
// Added types
type Employee struct {
	Id   int
	Name string
}

// Added manager subtype
type Manager struct {
	Employee
}

func (manager Manager) getId() int {
	return manager.Id
}

func (manager Manager) getName() string {
	return manager.Name
}

// Subclasses implement salary interface
func (manager Manager) calculateSalary() float64 {
	return float64(manager.Id * 5000)
}

// Added developer subtype
type Developer struct {
	Employee
}

func (developer Developer) getId() int {
	return developer.Id
}

func (developer Developer) getName() string {
	return developer.Name
}

// Subclasses implement salary interface
func (developer Developer) calculateSalary() float64 {
	return float64(developer.Id * 3000)
}

// Added intern subtype
type Intern struct {
	Employee
}

func (intern Intern) getId() int {
	return intern.Id
}

func (intern Intern) getName() string {
	return intern.Name
}

// Subclasses implement salary interface
func (intern Intern) calculateSalary() float64 {
	return float64(intern.Id * 1000)
}

// Removed SalaryCalculator because it was redundant. In a more complex scenario it would be useful to outsource the calculation to a separate class.
func main() {
	var manager EmployeeInterface
	manager = Manager{Employee{Id: 1, Name: "John Doe"}}

	var developer EmployeeInterface
	developer = Developer{Employee{Id: 2, Name: "Jane Smith"}}

	var intern EmployeeInterface
	intern = Intern{Employee{Id: 3, Name: "Mike Johnson"}}

	var salaries []float64

	salaries = append(salaries, manager.(SalaryInterface).calculateSalary())
	salaries = append(salaries, developer.(SalaryInterface).calculateSalary())
	salaries = append(salaries, intern.(SalaryInterface).calculateSalary())

	fmt.Println(salaries)
}

File: salary.php
<?php

// Employee class represents an employee in a company
class Employee {
    private $id;
    private $name;
    private $position;

    public function __construct($id, $name, $position) {
        $this->id = $id;
        $this->name = $name;
        $this->position = $position;
    }

    public function getId() {
        return $this->id;
    }

    public function getName() {
        return $this->name;
    }

    public function getPosition() {
        return $this->position;
    }
}

// SalaryCalculator class calculates the salary of employees
class SalaryCalculator {
    private $employees = [];

    public function addEmployee(Employee $employee) {
        $this->employees[] = $employee;
    }

    public function calculateSalary() {
        $salaries = [];
        foreach ($this->employees as $employee) {
            if ($employee->getPosition() === 'Manager') {
                $salaries[] = $this->calculateManagerSalary($employee);
            } elseif ($employee->getPosition() === 'Developer') {
                $salaries[] = $this->calculateDeveloperSalary($employee);
            } else {
                $salaries[] = $this->calculateInternSalary($employee);
            }
        }
        return $salaries;
    }

    private function calculateManagerSalary($employee) {
        // Complex calculation for manager salary
        return $employee->getId() * 5000;
    }

    private function calculateDeveloperSalary($employee) {
        // Complex calculation for developer salary
        return $employee->getId() * 3000;
    }

    private function calculateInternSalary($employee) {
        // Complex calculation for intern salary
        return $employee->getId() * 1000;
    }
}

$employee1 = new Employee(1, 'John Doe', 'Manager');
$employee2 = new Employee(2, 'Jane Smith', 'Developer');
$employee3 = new Employee(3, 'Mike Johnson', 'Intern');

$calculator = new SalaryCalculator();
$calculator->addEmployee($employee1);
$calculator->addEmployee($employee2);
$calculator->addEmployee($employee3);

$salaries = $calculator->calculateSalary();
print_r($salaries);
